pkg/reconciler/taskrun/taskrun.go

```diff
@@ -465,67 +470,33 @@ func (c *Reconciler) handlePodCreationError(tr *v1alpha1.TaskRun, err error) {
-			}
-		}
-	}
-	return nil
-}
+// failTaskRun stops a TaskRun with the provided Reason
+// If a pod is associated to the TaskRun, it stops it
+// failTaskRun function may return an error in case the pod could not be deleted
+// failTaskRun may update the local TaskRun status, but it won't push the updates to etcd
+func (c *Reconciler) failTaskRun(tr *v1alpha1.TaskRun, reason, message string) error {
 
-func getResults(results []v1alpha1.PipelineResourceResult) ([]v1alpha1.TaskRunResult, []v1alpha1.PipelineResourceResult) {
-	var taskResults []v1alpha1.TaskRunResult
-	var pipelineResourceResults []v1alpha1.PipelineResourceResult
-	for _, r := range results {
-		switch r.ResultType {
-		case v1alpha1.TaskRunResultType:
-			taskRunResult := v1alpha1.TaskRunResult{
-				Name:  r.Key,
-				Value: r.Value,
-			}
-			taskResults = append(taskResults, taskRunResult)
-		case v1alpha1.PipelineResourceResultType:
-			fallthrough
-		default:
-			pipelineResourceResults = append(pipelineResourceResults, r)
-		}
-	}
-	return taskResults, pipelineResourceResults
-}
+	c.Logger.Warn("stopping task run %q because of %q", tr.Name, reason)
+	tr.Status.MarkResourceFailed(reason, errors.New(message))
 
-func (c *Reconciler) updateStatus(taskrun *v1alpha1.TaskRun) (*v1alpha1.TaskRun, error) {
-	newtaskrun, err := c.taskRunLister.TaskRuns(taskrun.Namespace).Get(taskrun.Name)
-	if err != nil {
-		return nil, fmt.Errorf("error getting TaskRun %s when updating status: %w", taskrun.Name, err)
-	}
-	if !reflect.DeepEqual(taskrun.Status, newtaskrun.Status) {
-		newtaskrun.Status = taskrun.Status
-		return c.PipelineClientSet.TektonV1alpha1().TaskRuns(taskrun.Namespace).UpdateStatus(newtaskrun)
-	}
-	return newtaskrun, nil
-}
+	// update tr completed time
+	tr.Status.CompletionTime = &metav1.Time{Time: time.Now()}
 
-func (c *Reconciler) updateLabelsAndAnnotations(tr *v1alpha1.TaskRun) (*v1alpha1.TaskRun, error) {
-	newTr, err := c.taskRunLister.TaskRuns(tr.Namespace).Get(tr.Name)
-	if err != nil {
-		return nil, fmt.Errorf("error getting TaskRun %s when updating labels/annotations: %w", tr.Name, err)
+	if tr.Status.PodName == "" {
+		c.Logger.Warnf("task run %q has no pod running yet", tr.Name)
+		return nil
 	}
-	if !reflect.DeepEqual(tr.ObjectMeta.Labels, newTr.ObjectMeta.Labels) || !reflect.DeepEqual(tr.ObjectMeta.Annotations, newTr.ObjectMeta.Annotations) {
-		newTr.ObjectMeta.Labels = tr.ObjectMeta.Labels
-		newTr.ObjectMeta.Annotations = tr.ObjectMeta.Annotations
-		return c.PipelineClientSet.TektonV1alpha1().TaskRuns(tr.Namespace).Update(newTr)
+
+	// tr.Status.PodName will be empty if the pod was never successfully created. This condition
+	// can be reached, for example, by the pod never being schedulable due to limits imposed by
+	// a namespace's ResourceQuota.
+	err := c.KubeClientSet.CoreV1().Pods(tr.Namespace).Delete(tr.Status.PodName, &metav1.DeleteOptions{})
```

Comment:
I am wondering if some kind of if statement should have been placed here to guard against scenarios when a TaskRun reason is `v1beta1.TaskRunReasonTimedOut` or `v1beta1.TaskRunCancelled`. As of now, this is deleting pods for TaskRuns that have timed out or been cancelled, which amongst other things, deletes the TaskRun logs.